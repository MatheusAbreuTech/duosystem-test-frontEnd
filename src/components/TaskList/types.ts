export type Filters = 'all' | 'status' | 'createdDate'

export type DescriptionStylesProps = {
  taskisfinished: boolean
}
