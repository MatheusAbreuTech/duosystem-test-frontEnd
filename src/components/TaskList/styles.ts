import styled from 'styled-components'
import { DescriptionStylesProps } from './types'

export const Wrapper = styled.div`
  width: 100%;
`

export const List = styled.ul`
  width: 100%;
  list-style: none;
`

export const Item = styled.li`
  height: auto;
  width: 100%;
  border-left: 0.3rem solid #00ecf1;
  display: flex;
  align-items: center;
  justify-content: space-between;
  background: #fff;
  border-radius: 0.5rem;
  padding: 1rem;
  margin: 1rem 0;
  gap: 3rem;
  box-shadow: 0px 0px 20px 0px rgba(132, 132, 132, 0.2);

  @media (max-width: 700px) {
    flex-direction: column;
    margin: 2rem 0;
  }
`

export const Description = styled.p<DescriptionStylesProps>`
  font-size: 1.8rem;
  font-style: ${(props) => (props.taskisfinished ? 'italic' : 'normal')};
  text-decoration: ${(props) =>
    props.taskisfinished ? 'line-through' : 'none'};
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 300px;
  text-align: start;

  @media (max-width: 700px) {
    text-align: center;
  }
`

export const InfosWrapper = styled.div`
  max-width: 400px;
  width: 100%;
  display: flex;
  align-items: center;
  gap: 1rem;

  > button {
    width: 100%;
  }

  @media (max-width: 700px) {
    flex-direction: column;
  }
`

export const FiltersWrapper = styled.div`
  max-width: 70%;
  height: auto;
  display: flex;
  align-items: center;
  gap: 1rem;

  > span {
    font-size: 1.6rem;
  }

  @media (max-width: 700px) {
    max-width: 100%;
  }
`
