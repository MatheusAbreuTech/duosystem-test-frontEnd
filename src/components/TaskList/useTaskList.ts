import useTaskStore from '@/store/zustand'
import { Task } from '@/types/Task'
import { useEffect, useState } from 'react'
import { FilterDirection, Filters } from './types'

export const useTaskList = () => {
  const {
    taskList,
    changeTaskStatusById,
    searchValue,
    removeTaskById,
    searchTasksByDescription
  } = useTaskStore()
  const [editTaskId, setEditTaskId] = useState('')
  const [data, setData] = useState<Task[]>(taskList)
  const [filterBy, setFilterBy] = useState<Filters>('all')
  const [filterDirection, setFilterDirection] = useState<FilterDirection>('asc')

  useEffect(() => {
    let copyData = [...taskList]

    if (filterBy === 'status') {
      copyData = copyData.filter((task) => task.status)
    } else if (filterBy === 'createdDate') {
      copyData = copyData.sort(
        (a, b) =>
          (filterDirection === 'asc' ? 1 : -1) *
          (b.createDate.getTime() - a.createDate.getTime())
      )
    }

    copyData = copyData.filter((task) =>
      task.description.toLowerCase().includes(searchValue.toLowerCase())
    )

    setData(copyData)
  }, [searchValue, filterBy, taskList, filterDirection])

  const handleFilterClick = (filter: Filters) => {
    if (filter === filterBy) {
      setFilterDirection(filterDirection === 'asc' ? 'desc' : 'asc')
    } else {
      setFilterBy(filter)
      setFilterDirection('asc')
    }

    if (filter === 'all') {
      searchTasksByDescription('')
      setData(taskList)
    }
  }

  return {
    data,
    editTaskId,
    filterBy,
    filterDirection,
    handleFilterClick,
    setEditTaskId,
    changeTaskStatusById,
    removeTaskById
  }
}
