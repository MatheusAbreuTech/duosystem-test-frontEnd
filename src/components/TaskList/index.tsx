'use client'

import Button from '../Button'
import * as S from './styles'
import EditTask from '../EditTask'
import { useTaskList } from './useTaskList'
import useTaskStore from '@/store/zustand'

const TaskList = () => {
  const { data, editTaskId, setEditTaskId, handleFilterClick } = useTaskList()
  const { changeTaskStatusById, removeTaskById } = useTaskStore()

  return (
    <S.Wrapper>
      <S.FiltersWrapper>
        <span>Filtrar por: </span>
        <Button
          color="transparent"
          onClick={() => handleFilterClick('all')}
          data-testId="BtnFilterByAll"
        >
          Todas
        </Button>
        <Button
          color="transparent"
          onClick={() => handleFilterClick('status')}
          data-testId="BtnFilterByStatus"
        >
          Concluidas
        </Button>
        <Button
          color="transparent"
          onClick={() => handleFilterClick('createdDate')}
          data-testId="BtnFilterCreatedDate"
        >
          Data de criação
        </Button>
      </S.FiltersWrapper>

      {data.length > 0 ? (
        <S.List data-testId="list-data">
          {data.map((task) => {
            const formattedDate = new Date(task.createDate).toLocaleString(
              'pt-BR'
            )
            return (
              <S.Item key={task.id}>
                {editTaskId === task.id ? (
                  <EditTask taskId={task.id} setEditTaskId={setEditTaskId} />
                ) : (
                  <S.Description taskisfinished={task.status}>
                    {task.description}
                  </S.Description>
                )}

                <S.InfosWrapper>
                  <p>{formattedDate}</p>
                  <Button
                    color="green"
                    onClick={() => changeTaskStatusById(task.id)}
                    data-testId={`changeStatusButton_${task.id}`}
                  >
                    ${task.status ? 'Desmarcar' : 'Concluir'}
                  </Button>
                  {editTaskId !== task.id && (
                    <Button
                      color="darkBlue"
                      onClick={() => setEditTaskId(task.id)}
                      data-testId={`editButton_${task.id}`}
                    >
                      Editar
                    </Button>
                  )}
                  <Button
                    color="red"
                    onClick={() => removeTaskById(task.id)}
                    data-testId={`removeTaskButton_${task.id}`}
                  >
                    Excluir
                  </Button>
                </S.InfosWrapper>
              </S.Item>
            )
          })}
        </S.List>
      ) : (
        <p>Nenhuma tarefa encontrada</p>
      )}
    </S.Wrapper>
  )
}

export default TaskList
