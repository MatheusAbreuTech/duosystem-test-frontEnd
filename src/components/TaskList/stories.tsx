import { Meta, StoryObj } from '@storybook/react'
import TaskList from '.'
import useTaskStore from '@/store/zustand'
import Button from '../Button'

export default {
  title: 'TaskList',
  component: TaskList
} as Meta

const Template = () => {
  const { addTask } = useTaskStore()

  return (
    <>
      <Button
        color="darkBlue"
        onClick={() =>
          addTask(`descrição da tarefa ${Math.floor(Math.random() * 100)}`)
        }
      >
        Adicionar tarefa
      </Button>
      <TaskList />
    </>
  )
}

export const Default: StoryObj = {}

export const TaskListWithTasks = Template.bind({})
