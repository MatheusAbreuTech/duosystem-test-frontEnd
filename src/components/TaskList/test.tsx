import { fireEvent, render } from '@testing-library/react'
import TaskList from '../TaskList'
import { useTaskList } from './useTaskList'
import { act, renderHook } from '@testing-library/react-hooks'
import useTaskStore from '@/store/zustand'

jest.mock('@/store/zustand')
const mockedUseTaskStore = useTaskStore as jest.MockedFunction<
  typeof useTaskStore
>

describe('<TaskList />', () => {
  it('should correctly render the component', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: []
    })
    const { container } = render(<TaskList />)
    expect(container).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should only render a message when it has no task', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: [],
      searchValue: ''
    })

    const { getByText, queryByTestId } = render(<TaskList />)
    expect(getByText('Nenhuma tarefa encontrada')).toBeInTheDocument()
    expect(queryByTestId('list-data')).not.toBeInTheDocument()
  })

  it('should render tasks when provided', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: [
        {
          id: 'testId',
          description: 'Test',
          createDate: new Date('2022-01-01'),
          status: false
        }
      ],
      searchValue: ''
    })
    const { getByText } = render(<TaskList />)
    expect(getByText('Test')).toBeInTheDocument()
  })

  it('should display the EditTask component when the edit button is clicked', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: [
        {
          id: 'testId',
          description: 'Test',
          createDate: new Date('2022-01-01'),
          status: false
        }
      ],
      searchValue: ''
    })

    const { queryByTestId, getByTestId } = render(<TaskList />)

    const button = getByTestId(`editButton_testId`)

    fireEvent.click(button)

    expect(queryByTestId('editTaskComponent_testId')).toBeInTheDocument()
    expect(button).not.toBeInTheDocument()
  })

  it('should call the changeTaskStatusById function when clicking the change status button', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: [
        {
          id: 'testId',
          description: 'Test',
          createDate: new Date('2022-01-01'),
          status: false
        }
      ],
      searchValue: '',
      changeTaskStatusById: jest.fn()
    })

    const { getByTestId } = render(<TaskList />)

    const changeStatusButton = getByTestId('changeStatusButton_testId')

    fireEvent.click(changeStatusButton)

    expect(useTaskStore().changeTaskStatusById).toHaveBeenCalledWith('testId')
  })

  it('should call the removeTaskById function when clicking the remove task button', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: [
        {
          id: 'testId',
          description: 'Test',
          createDate: new Date('2022-01-01'),
          status: false
        }
      ],
      searchValue: '',
      removeTaskById: jest.fn()
    })

    const { getByTestId } = render(<TaskList />)

    const removeButton = getByTestId('removeTaskButton_testId')

    fireEvent.click(removeButton)

    expect(useTaskStore().removeTaskById).toHaveBeenCalledWith('testId')
  })
})

describe('useTaskList', () => {
  it('should start with the correct initial values', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: [],
      searchValue: ''
    })
    const { result } = renderHook(() => useTaskList())
    expect(result.current.data).toEqual([])
    expect(result.current.editTaskId).toBe('')
  })

  it('should return the list of tasks according to the filters passed when calling the handleFilterClick function', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: [
        {
          id: 'testId',
          description: 'Test',
          createDate: new Date('2022-01-01'),
          status: false
        },
        {
          id: 'testId2',
          description: 'Test',
          createDate: new Date('2021-12-31'),
          status: true
        }
      ],
      searchValue: '',
      searchTasksByDescription: jest.fn()
    })
    const { result } = renderHook(() => useTaskList())

    expect(result.current.data).toHaveLength(2)

    act(() => {
      result.current.handleFilterClick('all')
    })

    expect(result.current.data).toHaveLength(2)

    act(() => {
      result.current.handleFilterClick('status')
    })

    expect(result.current.data).toHaveLength(1)
    expect(result.current.data).toEqual([
      {
        id: 'testId2',
        description: 'Test',
        createDate: new Date('2021-12-31'),
        status: true
      }
    ])

    act(() => {
      result.current.handleFilterClick('createdDate')
    })

    const dataFiltered = result.current.data.sort(
      (a, b) => b.createDate.getTime() - a.createDate.getTime()
    )

    expect(result.current.data).toEqual(dataFiltered)
  })

  it('should filter the list of tasks if the searchValue value is already defined', () => {
    mockedUseTaskStore.mockReturnValue({
      taskList: [
        {
          id: 'testId',
          description: 'Test',
          createDate: new Date('2022-01-01'),
          status: false
        },
        {
          id: 'testId2',
          description: 'Test2',
          createDate: new Date('2022-01-02'),
          status: false
        },
        {
          id: 'testId3',
          description: 'Test3',
          createDate: new Date('2022-01-03'),
          status: false
        },
        {
          id: 'testId4',
          description: 'Test4',
          createDate: new Date('2022-01-04'),
          status: false
        }
      ],
      searchValue: 'Test4'
    })

    const { result } = renderHook(() => useTaskList())

    expect(result.current.data).toEqual([
      {
        id: 'testId4',
        description: 'Test4',
        createDate: new Date('2022-01-04'),
        status: false
      }
    ])
  })
})
