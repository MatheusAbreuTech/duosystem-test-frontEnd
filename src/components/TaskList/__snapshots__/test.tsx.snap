// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskList /> should correctly render the component 1`] = `
.c2 {
  background-color: transparent;
  color: #0c2e73;
  border: none;
  border-radius: 0.5rem;
  padding: 1rem;
  max-height: 4.2rem;
  height: 100%;
  cursor: pointer;
  font-size: 1.6rem;
}

.c0 {
  width: 100%;
}

.c1 {
  max-width: 70%;
  height: auto;
  display: flex;
  align-items: center;
  gap: 1rem;
}

.c1 >span {
  font-size: 1.6rem;
}

@media (max-width: 700px) {
  .c1 {
    max-width: 100%;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span>
      Filtrar por: 
    </span>
    <button
      class="c2"
      color="transparent"
      data-testid="BtnFilterByAll"
    >
      Todas
    </button>
    <button
      class="c2"
      color="transparent"
      data-testid="BtnFilterByStatus"
    >
      Concluidas
    </button>
    <button
      class="c2"
      color="transparent"
      data-testid="BtnFilterCreatedDate"
    >
      Data de criação
    </button>
  </div>
  <p>
    Nenhuma tarefa encontrada
  </p>
</div>
`;
